### PR TITLE
Encrypt metadata

### PR DESCRIPTION
**Pull request by nfarah86**
wip- encrypting metadata

### CODE REVIEW

File: pkg/storage/streams/store.go [hunk 1]
```diff
@@ -38,12 +38,29 @@ type Meta struct {
...
+	logger, _ := zap.NewDevelopment()
+	defer logger.Sync()
+
+	logger.Debug("we are in streamstore convertmeta-1")
+	fmt.Println("lastSegmentMeta in convertMeta streamstore: ", lastSegmentMeta, "\n\n\n\n\n")
+
+
 	msi := pb.MetaStreamInfo{}
+	
+	// THIS IS THE ERROR THAT NEEDS TO GET FIXED - DECRYPT BEFORE UNMARSHALING
```
Comment (dylanlott) [inline pkg/storage/streams/store.go]: Nit: Mark this with TODO for maybe easier searching

File: pkg/storage/streams/store.go [hunk 2]
```diff
@@ -38,12 +38,29 @@ type Meta struct {
...
+	// -- SEE META
+
 	err := proto.Unmarshal(lastSegmentMeta.Data, &msi)
+	fmt.Println("This is msi- 1", msi)
+	logger.Debug("we are in streamstore convertmeta-2")
+	fmt.Println("this is msi after marshaling convertMeta streamstore",msi, "\n\n\n\n\n")
 	if err != nil {
+		logger.Debug("we are in streamstore convertmeta-3")
+		fmt.Println("this is meta: ", msi)
```
Comment (dylanlott) [inline pkg/storage/streams/store.go]: Nit: remove fmt comments

File: pkg/storage/streams/store.go [hunk 3]
```diff
@@ -187,32 +213,53 @@ func (s *streamStore) Put(ctx context.Context, path paths.Path, data io.Reader,
...
+			
 			lastSegmentMeta, err := proto.Marshal(&msi)
 			if err != nil {
 				return nil, nil, err
 			}
-			return lastSegmentPath, lastSegmentMeta, nil
-		})
+
+			logger.Debug("WE ARE IN  put STREAMSTORE -1\n\n\n\n\n\n")
+			fmt.Println("this is the marshalled last segmentMeta: ", lastSegmentMeta, "\n\n\n")
```
Comment (dylanlott) [inline pkg/storage/streams/store.go]: Nit: remove comments

File: pkg/storage/streams/store.go [hunk 4]
```diff
@@ -187,32 +213,53 @@ func (s *streamStore) Put(ctx context.Context, path paths.Path, data io.Reader,
...
 
+	logger.Debug("WE ARE IN STREAMSTORE -7 \n\n\n\n\n\n")
 	resultMeta := Meta{
 		Modified:   putMeta.Modified,
 		Expiration: expiration,
 		Size:       streamSize,
 		Data:       metadata,
 	}
 
+	logger.Debug("WE ARE IN STREAMSTORE -8 \n\n\n\n\n\n")
```
Comment (dylanlott) [inline pkg/storage/streams/store.go]: Nit: Probably get rid of these, too

File: pkg/storage/streams/store.go (hunk 3 above)
Comment (nfarah86) [reply pkg/storage/streams/store.go]: yup! it's a WIP- not ready yet (just finished working on a bug) should be ready now

**Comment by nfarah86** [comment]
@kaloyan-raev can you review please?

File: pkg/storage/streams/store.go [hunk 5]
```diff
@@ -232,18 +242,33 @@ func (s *streamStore) Get(ctx context.Context, path paths.Path) (rr ranger.Range
...
 	}
 
-	derivedKey, err := path.DeriveContentKey(s.rootKey)
+	msi := pb.MetaStreamInfo{}
+	err = proto.Unmarshal(decryptedMetaData, &msi)
+	if err != nil {
+		return nil, Meta{}, err
+	}
+
+	lastlastSegMetaDecrypted := segments.Meta{
```
[the remark this answers is not in the document]
Comment (nfarah86) [reply pkg/storage/streams/store.go]: yup typo good catch

File: pkg/storage/streams/store.go [hunk 6]
```diff
@@ -626,3 +671,30 @@ func getEncryptedKeyAndNonce(m *pb.SegmentMeta) ([]byte, *eestream.Nonce) {
...
+
+	e, err := cipher.Decrypt(encryptedKey, derivedKey, keyNonce)
+	if err != nil {
+		return nil, err
+	}
+
+	var encKey eestream.Key
+	copy(encKey[:], e)
+
+	decryptedStreamInfo, err = cipher.Decrypt(streamMeta.EncryptedStreamInfo, &encKey, &nonce)
```
Comment (mobyvb) [inline pkg/storage/streams/store.go]: Where does this `&nonce` get set? It is never initialized, so it is probably not the same nonce that the stream info was encrypted with

File: pkg/storage/streams/store.go [hunk 7]
```diff
@@ -222,38 +210,37 @@ func (s *streamStore) Put(ctx context.Context, path paths.Path, data io.Reader,
 			if err != nil {
 				return nil, nil, err
 			}
-
+			encryptedStreamInfo, err := cipher.Encrypt(streamInfo, &encKey, &nonce)
```
Comment (mobyvb) [inline pkg/storage/streams/store.go]: How is this nonce being generated? Is it being generated in the same way when we decrypt stream info?

File: pkg/storage/streams/store.go [hunk 8]
```diff
@@ -279,26 +266,39 @@ func (s *streamStore) Get(ctx context.Context, path paths.Path) (rr ranger.Range
...
 	}
-
 	lastSegmentRanger, lastSegmentMeta, err := s.segments.Get(ctx, encPath.Prepend("l"))
 	if err != nil {
 		return nil, Meta{}, err
 	}
 
-	streamMeta := pb.StreamMeta{}
-	err = proto.Unmarshal(lastSegmentMeta.Data, &streamMeta)
+	derivedKey, err := path.DeriveContentKey(s.rootKey)
```
Comment (mobyvb) [inline pkg/storage/streams/store.go]: You have this code almost everywhere before you call `getDecryptedStreamInfo()`

```
derivedKey, err := path.DeriveContentKey(s.rootKey)
if err != nil {
	return nil, Meta{}, err
}
key := (*eestream.Key)(derivedKey)
```
I think you should change `getDecryptedStreamInfo` to take in a path and the rootKey, and you move this repeated logic inside that function. That way you can get rid of a lot of extra code

Comment (nfarah86) [reply pkg/storage/streams/store.go]: yup i can do that!

**Comment by nfarah86** [comment]
yup @navillasa convertMeta function was slightly changed- it had to do with unmarshaling data. All fixed :)

**Pull request merged**